python/pyspark/sql/tests/pandas/test_pandas_transform_with_state.py

```diff
@@ -336,6 +341,263 @@ def check_results(batch_df, batch_id):
+                # because this is a new timer registered in batch1 and
+                # different from the one registered in batch 0
+                current_batch_expired_timestamp = \
+                    batch_df.filter(batch_df["countAsString"] == -1).first()['timeValues']
+                assert(current_batch_expired_timestamp > self.first_expired_timestamp)
+
+        self._test_transform_with_state_in_pandas_proc_timer(ProcTimeStatefulProcessor(), check_results)
+
+    def _test_transform_with_state_in_pandas_event_time(self, stateful_processor, check_results):
+        import pyspark.sql.functions as f
+
+        input_path = tempfile.mkdtemp()
+
+        def prepare_batch1(input_path):
+            with open(input_path + "/text-test3.txt", "w") as fw:
+                fw.write("a, 20\n")
+
+        def prepare_batch2(input_path):
+            with open(input_path + "/text-test1.txt", "w") as fw:
+                fw.write("a, 4\n")
+
+        def prepare_batch3(input_path):
+            with open(input_path + "/text-test2.txt", "w") as fw:
+                fw.write("a, 11\n")
+                fw.write("a, 13\n")
+                fw.write("a, 15\n")
+
+        prepare_batch1(input_path)
+        prepare_batch2(input_path)
+        prepare_batch3(input_path)
+
+        df = self._build_test_df(input_path)
+        df = df.select("id",
+                       f.from_unixtime(f.col("temperature")).alias("eventTime").cast("timestamp")) \
+            .withWatermark("eventTime", "10 seconds")
+
+        for q in self.spark.streams.active:
+            q.stop()
+        self.assertTrue(df.isStreaming)
+
+        output_schema = StructType(
+            [
+                StructField("id", StringType(), True),
+                StructField("timestamp", StringType(), True)
+            ]
+        )
+
+        query_name = "event_time_test_query"
+        q = (
+            df.groupBy("id")
+            .transformWithStateInPandas(
+                statefulProcessor=stateful_processor,
+                outputStructType=output_schema,
+                outputMode="Update",
+                timeMode="eventtime",
+            )
+            .writeStream.queryName(query_name)
+            .foreachBatch(check_results)
+            .outputMode("update")
+            .start()
+        )
+
+        self.assertEqual(q.name, query_name)
+        self.assertTrue(q.isActive)
+        q.processAllAvailable()
+        q.awaitTermination(10)
+        self.assertTrue(q.exception() is None)
+
+    def test_transform_with_state_in_pandas_event_time(self):
+        def check_results(batch_df, batch_id):
+            if batch_id == 0:
+                assert set(batch_df.sort("id").collect()) == {
+                    Row(id="a", timestamp="20")
+                }
+            elif batch_id == 1:
+                # event time = 4 will be discarded because the watermark = 15 - 10 = 5
```

Comment:
This is not true - watermark for eviction is 5 but watermark for late record is 0, hence ("a", 4) is not dropped. This is exactly the reason you still see event for "a". Otherwise you shouldn't have ("a", 20).

Reply:
You might wonder how this works differently with Scala tests - AddData() & CheckNewAnswer() will trigger no-data batch, hence executing two batches.

Reply:
Thanks for leaving the comments! By reading your comments I realized I did not quite understand the difference between watermark for eviction and watermark for late record before. 
The test case should be still fine, I just deleted the comments. Dropping late record will be tested more throughly in the chaining of operator PR.